### ISSUE TITLE
inner and outer functions for complex

### ISSUE THREAD

**Issue by stevengj** [gridap/Gridap.jl#433]
Currently, https://github.com/gridap/Gridap.jl/blob/master/src/TensorValues/Operations.jl defines
```jl
inner(a::Real,b::Real) = a*b
outer(a::Real,b::Real) = a*b
```
which led to a `MethodError` for `outer` when we tried to assemble an `FESource` for a complex-valued function (see also #411).

Can these be changed to `Number` instead of `Real`?  e.g. to
```jl
inner(a::Number,b::Number) = a'*b
outer(a::Number,b::Number) = a*b'
```
and presumably there are other methods that need to be generalized from `Real` to `Number` too.

**Comment by stevengj** [comment]
(Normally, one would define these with complex conjugations, but I'm not entirely clear on whether matters are different for how they are used in this FEM context.)

**Comment by stevengj** [comment]
In our complex-PDE adjoint code, we found that we needed to define the unconjugated product `Gridap.outer(a::Number,b::Number) = a*b` for it to work.  Presumably this is because there are no conjugations elsewhere in Gridap and once needs to be consistent.

This is fine with me.    Any conjugations that we need can be specified explicitly in our `FESource` terms or elsewhere, and arguably it is clearer to leave this up to the user.

(For example, dissipative wave equations like Maxwell's equations and the scalar Helmholtz equations typically lead to weak forms that are complex-symmetric rather than Hermitian, and unconjugated "inner products" are common — this is [reciprocity](https://en.wikipedia.org/wiki/Reciprocity_(electromagnetism)).)

**Comment by fverdugo** [comment]
Good point. In any case, one can use user-defined products to define the FESource.

```julia
myprod(a,b) = a*b
f = 2
FESource( v->operate(myprod,v,f), trian, quad )
```
But, I agree,  it is necessary to find some "sensible" defaults when working with complex numbers.

**Comment by stevengj** [comment]
In general, I would define methods to work with `Number` (rather than `Real`) as much as possible.   There is no mathematical reason why one couldn't have PDEs with quaternion functions, dual numbers, etcetera, as well as complex numbers.

**Comment by stevengj** [comment]
Looks like this was fixed for `outer` by a24a2db1b26e18bae125c3da633ede784d57a6a0.

For `inner`, you currently throw an error for `Complex` values, but I think you want to delete [that method](https://github.com/gridap/Gridap.jl/blob/15c0130d762454e0e96ec94da3c7343a57256722/src/TensorValues/Operations.jl#L218-L220) and use the unconjugated product.

**Comment by fverdugo** [comment]
fixed in PR #492
